Feat/v3

PatrickAlphaC opened the pull request:
work in progress to solve: https://github.com/security-alliance/safe-harbor/issues/39

TODO:
- Get final list of owner addresses and chains to deploy new contracts to

TODO (after deployment): 
- Update references to on-chain contracts after deployment in legal docs

PatrickAlphaC committed:
wip

PatrickAlphaC committed:
feat: refactored code to fix event emissions, errors, gas-intensive loops, and removed fallback registry

PatrickAlphaC committed:
feat: aderyn fixes

PatrickAlphaC committed:
fix: solana submodule

PatrickAlphaC committed:
feat: chain validation outsourced

PatrickAlphaC committed:
feat: initializer function using legacy registry

PatrickAlphaC committed:
feat: included other URIs for the agreement URI

PatrickAlphaC committed:
feat: updated legal docs for other valid URIs

Robert-MacWha commented:
Is there a reason for the libs to be copied instead of references for legacy?  Might have just been a mistake when cloning?

PatrickAlphaC committed:
feat: tests and scripts

PatrickAlphaC committed:
feat: updated names in config

PatrickAlphaC committed:
feat: found a foundry bug, changing tests to accomodate

PatrickAlphaC committed:
feat: removed lib from legacy

Robert-MacWha commented:
The constructor is clever.  If we want to retain the same address, we can just include all the adopters and the invalid ones will be skipped.  And if there was no deployed registry it'll skip over the entire block.

Would it be better to exclude the chain validation logic from the registry?  Right now the registry serves dual responsibilities, and opens the door to needing a new registry (or at least deprecating the current one) if the interface for agreement verification changes too much.

I would tentatively suggest:
```solidity
contract SafeHarborRegistry is IRegistry { // Remove Ownable since no onlyOwner methods 
    error SafeHarborRegistry__NoAgreement();
    error SafeHarborRegistry__ZeroAddress();
    
    mapping(address entity => address details) private agreements;
    
    event SafeHarborAdoption(address indexed adopter, address agreementAddress);
    event LegacyDataMigrated(address indexed legacyRegistry, uint256 migratedCount);
    
    constructor(address _legacyRegistry, address[] memory _adopters)
    function adoptSafeHarbor(address _agreementAddress) external
    function version() external pure returns (string memory)
    function getAgreement(address _adopter) external view returns (address)
}

contract AgreementFactory {
    address private constant REGISTRY_ADDRESS;
    address private constant CHAIN_VALIDATOR_ADDRESS;
    function create(AgreementDetails memory details, address owner) external returns (address agreementAddress);
}
```

Or, if we want the chain validator to be variable, either using the `AgreementFactory` as the proxy or by letting the owners of an `Agreement` change their CHAIN_VALIDATOR_ADDRESS.  Unless there's a good reason for mutable validator addresses I'd make them constant, otherwise would vote to move it to the factory.

PatrickAlphaC committed:
feat: added create2 to agreements so they can be more different addresses

PatrickAlphaC committed:
feat: remove initializer

PatrickAlphaC committed:
feat: reintroduce via_ir :(

Robert-MacWha commented on registry-contracts/src/ChainValidator.sol:
```diff
@@ -0,0 +1,94 @@
+
+    /// @notice Internal function to remove a chain ID from the valid chains list.
+    /// @param _caip2ChainId The CAIP-2 chain ID to remove.
+    function _removeFromValidChainsList(string calldata _caip2ChainId) internal {
```
Unless I'm mistaken, isn't this the same n*m iteration as in the original code you raised?  Is it still a problem?

I was thinking, instead of storing a bool in the `validChains` mapping we could store a uint (index of the chainID in the list).  Then instead of iterating in `_removeFromValidChainsList` we could use that index directly, and update the index of the swapped chain.

Away from my desktop until tomorrow evening, but I can write the code then if this doesn't make sense.  Or if it's not an issue lmk.

PatrickAlphaC replied:
Good shout! Added in 935e9826303f500bfb637d0964d07247abdd8335

Robert-MacWha commented on registry-contracts/src/Agreement.sol:
```diff
@@ -0,0 +1,411 @@
+    /// @notice Removes multiple chains from the agreement by CAIP-2 IDs.
+    /// @param _caip2ChainIds Array of CAIP-2 IDs of the chains to remove
+    // aderyn-ignore-next-line(centralization-risk)
+    function removeChains(string[] memory _caip2ChainIds) external onlyOwner {
```
Are we also worried about this (or the below `removeAccounts` fn) from a gas perspective?  It's a similar n*m loop to that in `SafeHarborRegistryV2` that you originally highlighted?  If not that's fine. just checking if the situation is the same.

Robert-MacWha commented on registry-contracts/src/Agreement.sol:
```diff
@@ -0,0 +1,411 @@
+    mapping(string caip2ChainId => Account[]) private accounts;
+
+    // ----- EVENTS -----
+    event ProtocolNameUpdated(string newName);
```
```suggestion
    event ProtocolNameSet(string newName);
```

Robert-MacWha commented on registry-contracts/src/Agreement.sol:
```diff
@@ -0,0 +1,411 @@
+    // aderyn-ignore-next-line(centralization-risk)
+    function setBountyTerms(BountyTerms memory _bountyTerms) external onlyOwner {
+        _validateBountyTerms(_bountyTerms);
+        emit BountyTermsUpdated(_bountyTerms);
```
```suggestion
        emit BountyTermsSet(_bountyTerms);
```

PatrickAlphaC replied:
Fixed in 8bf0bca2f417576d10973a138ec095500ad00f94

DicksonWu654 committed:
feat: add AgreementFactory deployment to SafeHarbor script

DicksonWu654 committed:
refactor: update deployment script and agreement contracts to use ChainValidator interface

DicksonWu654 committed:
refactor: update Agreement and Factory tests to use ChainValidator instead of registry

DicksonWu654 committed:
refactor: clean up ChainValidatorTest formatting and improve readability

DicksonWu654 commented:
So far: 
- Add the deployment of the AgreementFactory inside of the deploy.s.sol
- Remove the Registry middleman function of calling isChainValid + removing the ownership from the registry
- Fix the tests because of the registry change
- Also move the tests for chain validator into it's own test file

DicksonWu654 commented:
TODO: Add V1 adoptors into the deployment script

DicksonWu654 committed:
refactor: update deployment script and helper configuration for improved address management and constants

DicksonWu654 committed:
feat: add BSC and ZKSYNC configurations and update adopters in HelperConfig

DicksonWu654 committed:
fix: update HelperConfig addresses for PARETO and SILO, and adjust adopters array size

DicksonWu654 committed:
refactor: update HelperConfig with new legacy registry addresses and streamline network configuration functions

DicksonWu654 commented:
> I think the final thing for us to do, would be the following.
> 
> In our `HelperConfig.s.sol` we have different configs for each chain:
> 
> ```solidity
> function getMainnetConfig() public pure returns (NetworkConfig memory) {
>         address[] memory adopters = new address[](5);
>         adopters[0] = RHEO;
>         adopters[1] = OPS_COVEFI_ETH;
>         adopters[2] = AAVE;
>         adopters[3] = IDLE_FINANCE;
>         adopters[4] = ENS;
>         return NetworkConfig({
>             owner: SEAL_MAINNET_OWNER,
>             legacyRegistry: DEFAULT_LEGACY_REGISTRY,
>             createx: CREATEX_ADDRESS,
>             adopters: adopters
>         });
>     }
> ```
> 
> I'd need someone to make a PR to go through each chain and add the correct config information. I think it might be easy enough to set the owner to be `SEAL_MAINNET_OWNER`. What do you think?

Alrighty done in my branch: https://github.com/PatrickAlphaC/safe-harbor/pull/1/commits/d1dc88034c024bde48c0317ae09553b4da9ca598

DicksonWu654 commented on registry-contracts/src/ChainValidator.sol:
```diff
@@ -0,0 +1,94 @@
+        for (uint256 i = 0; i < length; i++) {
+            validChains[_initialValidChains[i]] = true;
+            validChainsList.push(_initialValidChains[i]);
+            emit ChainValiditySet(_initialValidChains[i], true);
```
```suggestion
            if (!validChains[_initialValidChains[i]]) {
                validChains[_initialValidChains[i]] = true;
                validChainsList.push(_initialValidChains[i]);
                emit ChainValiditySet(_initialValidChains[i], true);
            }
```

A potential bug is we aren't checking for duplicates inside of the constructor. We are doing it inside of the setValidChains, but not here. Duplicate here = prob some unintended side effects later?

DicksonWu654 replied:
hehe trying to reduce the number of errors in the audit to 0

DicksonWu654 commented on registry-contracts/src/Agreement.sol:
```diff
@@ -0,0 +1,411 @@
+        _setDetails(_details);
+    }
+
+    // ----- USER-FACING STATE-CHANGING FUNCTIONS -----
```
Hmmm there's no setAgreementURI()... Should we add it?

Imo we should since they should be allowed to update the version of the legal contract which they're agreeing to

DicksonWu654 commented on registry-contracts/src/SafeHarborRegistry.sol:
```diff
@@ -0,0 +1,131 @@
+    /// @notice Function that records an adoption by msg.sender.
+    /// @param _agreementAddress The address of the agreement to adopt.
+    function adoptSafeHarbor(address _agreementAddress) external {
+        address adopter = msg.sender;
```
Should we have some address validation of making sure the address ain't 0?

I can't make a suggestion here cause I need to add an error hehe

PatrickAlphaC replied:
`msg.sender` is computationally infeasible to ever be 0.

DicksonWu654 replied:
Oh lol I meant the agreement address

Robert-MacWha replied:
Think it's fine.  The adopted address being 0 doesn't harm anything.

Robert-MacWha replied:
Actually semantically I think it'd mean the protocol is un-adopting safe harbor.

DicksonWu654 commented:
Nice to have features are scripts to adopt Safe HarborV3 from a json & a script to read it from the chain and print it out nicely... But nice to haves and maybe we can leave it for another PR

Robert-MacWha commented:
> Nice to have features are scripts to adopt Safe HarborV3 from a json & a script to read it from the chain and print it out nicely... But nice to haves and maybe we can leave it for another PR

Also those probably don't need to be present for the audit, since they're supplemental

DicksonWu654 commented:
Yeah exactly another PR

On Sun, Jan 4, 2026 at 4:32 PM Robert MacWha ***@***.***>
wrote:

> *Robert-MacWha* left a comment (security-alliance/safe-harbor#40)
> <https://github.com/security-alliance/safe-harbor/pull/40#issuecomment-3708444810>
>
> Nice to have features are scripts to adopt Safe HarborV3 from a json & a
> script to read it from the chain and print it out nicely... But nice to
> haves and maybe we can leave it for another PR
>
> Also those probably don't need to be present for the audit, since they're
> supplemental
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/security-alliance/safe-harbor/pull/40#issuecomment-3708444810>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIAWHKOPFEUZ7O3X3TYMBAT4FGBINAVCNFSM6AAAAACQLPMJUWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTOMBYGQ2DIOBRGA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

PatrickAlphaC committed:
feat: refactored for chainvaliidator to be behind a proxy

PatrickAlphaC committed:
Merge pull request #2 from PatrickAlphaC/feat/v3-dickson

Feat/v3 dickson

PatrickAlphaC committed:
feat: faster removal logic of chains using uint256

PatrickAlphaC committed:
feat: changed event to ProtocolNameSet and added tests

PatrickAlphaC committed:
feat: changed event name to BountyTermsSet

PatrickAlphaC committed:
feat: skip dups

PatrickAlphaC committed:
feat: added vanity miner

PatrickAlphaC committed:
feat: docs

DicksonWu654 commented:
True it’s a conviwnt way to set the adoption to 0

On Sun, Jan 4, 2026 at 10:18 PM Robert MacWha ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In registry-contracts/src/SafeHarborRegistry.sol
> <https://github.com/security-alliance/safe-harbor/pull/40#discussion_r2660184145>
> :
>
> > +    function adoptSafeHarbor(address _agreementAddress) external {
> +        address adopter = msg.sender;
>
> Think it's fine. The adopted address being 0 doesn't harm anything.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/security-alliance/safe-harbor/pull/40#discussion_r2660184145>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIAWHKMDKZIUC3N5X7J5JAD4FHJW3AVCNFSM6AAAAACQLPMJUWVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTMMRVGM4TAOBYGE>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

PatrickAlphaC committed:
Update registry-contracts/src/SafeHarborRegistry.sol

Co-authored-by: Robert MacWha <trebor.ahwcam@gmail.com>

PatrickAlphaC committed:
feat: fix lint, and fix nitpick

DicksonWu654 committed:
Audit fix: Missing bounty percentage and cap validation

feat: add bounty validation logic and corresponding tests

- Introduced new errors for bounty percentage exceeding maximum and aggregate bounty cap validation.
- Implemented validation logic in the _validateBountyTerms function.
- Added unit tests to cover scenarios for bounty percentage limits and aggregate cap conditions.

DicksonWu654 committed:
Audit Fix: Missing validation allows empty account addresses in agreement scope, creating ambiguous whitehat coverage

feat: add validation for empty account addresses in Agreement contract

- Introduced a new error for invalid account addresses.
- Implemented checks to ensure that account addresses are not empty during agreement creation and chain addition.
- Added unit tests to verify the behavior when empty account addresses are provided.

DicksonWu654 committed:
forge fmt

DicksonWu654 committed:
Audit fix: Missing contact details validation

feat: add validation for contact details in Agreement contract

- Introduced a new error for invalid contact details.
- Implemented validation to ensure contact names and contact fields are not empty during agreement creation and when setting contact details.
- Added unit tests to verify behavior for various invalid contact scenarios.

DicksonWu654 committed:
Audit fix: Free memory pointer corruption in inline assembly causes memory allocation panic and denial of service

Refactor AgreementFactory: Optimize assembly memory usage for salt generation

- Updated the assembly code to use a pointer for memory storage, improving efficiency in generating the final salt for agreement creation.
- Ensured that the memory pointer is updated correctly after usage to prevent potential memory issues.

DicksonWu654 committed:
Audit fix: Agreement::removeAccounts can leave chain with zero accounts

feat: add validation to prevent removal of all accounts in Agreement contract

- Introduced a new error to handle attempts to remove all accounts from a chain.
- Implemented checks in the removeAccounts function to ensure at least one account remains.
- Added unit tests to verify the new behavior when attempting to remove all accounts.

DicksonWu654 committed:
Audit Fix: Prefer calldata instead of memory for read-only external inputs

Refactor Agreement and ChainValidator contracts to use calldata for function parameters

- Updated function signatures in Agreement and ChainValidator contracts to use 'calldata' instead of 'memory' for string and array parameters, optimizing gas usage and improving efficiency.
- Functions affected include setProtocolName, setContactDetails, addChains, addAccounts in Agreement, and initialize in ChainValidator.

DicksonWu654 committed:
Audit Fix 2nd time: Missing validation allows empty account addresses in agreement scope, creating ambiguous whitehat coverage

feat: add validation for non-empty account addresses in Agreement contract

- Implemented checks to ensure that each account address provided in the addAccounts function is non-empty.
- Introduced a new error for invalid account addresses to handle cases where empty addresses are submitted.
- Added unit tests to verify that attempts to add accounts with empty addresses correctly revert the transaction.

DicksonWu654 committed:
Audit Fix: In Solidity don't initialize to default values
Refactor loop initialization in Agreement, ChainValidator, and SafeHarborRegistry contracts

- Updated for-loops to initialize the loop variable without an explicit starting value, improving code readability.
- Ensured consistency across multiple functions in the Agreement and ChainValidator contracts, as well as in the SafeHarborRegistry contract.

DicksonWu654 committed:
Audit FIx: Cache storage to prevent identical storage reads

Refactor Agreement contract: Optimize loop variable initialization

- Updated for-loops in the Agreement contract to store the length of arrays in a variable before iteration, enhancing code readability and performance.
- Ensured consistency in loop initialization across multiple functions within the contract.

DicksonWu654 committed:
Audit Fix: Remove obsolete final return statement when already using named returns
Refactor AgreementFactory: Remove redundant return statement in createAgreement function

- Eliminated the return statement in the createAgreement function as it was unnecessary, streamlining the code.
- This change enhances code clarity and maintains the function's intended behavior.

DicksonWu654 committed:
Audit Fix: Don't cache calldata array length

Refactor ChainValidator: Optimize loop variable usage in setValidChains and setInvalidChains functions

- Updated for-loops in the setValidChains and setInvalidChains functions to directly use the length of the input array, improving code clarity and reducing unnecessary variable declarations.
- This change enhances readability and maintains the intended functionality of the contract.

DicksonWu654 committed:
Audit fix: Inconsistencies in names and data types between legal agreement and smart contracts

Everything except for Version

Refactor Agreement contract: Update struct definitions and terminology for clarity

- Revised the definitions of DAO Adoption Procedures, BountyTerms, and IdentityRequirements structs to enhance clarity and consistency.
- Changed references from 'AgreementDetailsV2' to 'AgreementDetails' and updated variable names for better understanding.
- Ensured that all instances of struct definitions are aligned with the latest terminology and specifications.

DicksonWu654 committed:
Audit Fix: Optional signature field specified in legal agreement but not implemented in Chain struct

Refactor Agreement contract: Remove redundant signature reference and enhance clarity

- Eliminated the optional signature reference in the agreement section to streamline the document.
- Improved the overall clarity of the agreement by refining the language and structure of the contact details and bounty terms sections.

DicksonWu654 committed:
Version 3

DicksonWu654 committed:
Audit fix: Prefer calldata instead of memory for read-only external inputs - Fixed back to memory for gas

Refactor Agreement contract: Change addChains function parameter from calldata to memory

- Updated the addChains function to accept the _chains parameter as memory instead of calldata, improving compatibility with internal function calls.
- This change enhances the flexibility of the function while maintaining its intended functionality.

PatrickAlphaC committed:
feat: guarded salts

PatrickAlphaC committed:
Merge pull request #4 from PatrickAlphaC/v3-audit-remediation

V3 audit remediation

DicksonWu654 committed:
feat: add QUICKSWAP address to HelperConfig and update adopters array

- Introduced the QUICKSWAP address for the Polygon network in the HelperConfig contract.
- Expanded the adopters array in the getPolygonConfig function to include QUICKSWAP, enhancing the network configuration.

DicksonWu654 committed:
chore: update agreement document to version 3.0.0 and refine struct definitions

- Updated the version number in the agreement document to v3.0.0.
- Revised struct definitions for DAO Adoption Procedures, BountyTerms, and ContactDetails for improved clarity and consistency.
- Changed references to specific contract addresses and updated terminology to align with the latest specifications.

DicksonWu654 committed:
fix: correct typographical errors and enhance clarity in agreement document

- Fixed typographical errors in references to the Security Team and IdentityRequirements struct.
- Improved clarity in the language of the Conditions Precedent and Release by Protocol Community sections to ensure consistency and understanding.

DicksonWu654 committed:
generating agreement pdf, moving aderyn toml into registry-contracts, removing the redundant test.yml

DicksonWu654 committed:
docs: update acknowledgements in README.md to recognize contributions for V3 and Solana implementation

DicksonWu654 committed:
Merge remote-tracking branch 'upstream/main' into feat/v3

PatrickAlphaC commented:
Beautiful!
